src/Orchard/Tasks/Locking/Services/DistributedLockService.cs

```diff
@@ -0,0 +1,153 @@
+using System;
+using System.Data;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Autofac;
+using Orchard.Data;
+using Orchard.Environment;
+using Orchard.Exceptions;
+using Orchard.Logging;
+using Orchard.Services;
+using Orchard.Tasks.Locking.Records;
+
+namespace Orchard.Tasks.Locking.Services {
+
+    public class DistributedLockService : IDistributedLockService {
+        private readonly IMachineNameProvider _machineNameProvider;
+        private readonly ILifetimeScope _lifetimeScope;
+        private readonly IClock _clock;
+        private readonly object _semaphore = new object();
```

Comment:
@Piedone you are right but you shouldn't be. :) Let me explain.

Something I missed in my review yesterday: I believe this monitor is not needed at all, and @sfmskywalker also mentioned he thinks it's only needed from a unit test context. But give that we keep this monitor in place, it should not be held for this long. It should only be held for the duration of the atomic operation of trying to acquire the distributed lock, but if the current thread goes back to waiting because the distributed lock could not be acquired, then the monitor should be released while waiting, and then reacquired when the thread wakes up.

This will ensure that this monitor is only ever held extremely briefly I don't think a named lock should be necessary. So given this implementation you are right, but the implementation should not be this way in the first place.